### ISSUE TITLE
csv export for finding aid metadata

### ISSUE THREAD

**Issue by rlskoeser** [Princeton-CDH/mep-django#118]
Should include the following information for account holders with cards:
* [x] full normalized name
* [x] viaf if available 
* [x] earliest and latest transaction date on card
* [x] card images associated with the account
* [x] exclude partially known years when generating the date range

**Comment by rlskoeser** [comment]
Documenting feedback from @elspethgreen sent via email January 28:

> I've been slowly but surely working through the library export document (sylviabeach-cardholders) you initially sent in October.  I noticed initially that many of the cardholders were missing, and realized that they were not correctly designed as "has card" in the database.  This, in turn, revealed a lot of additional incorrectly-assigned borrowing events, which I've been correcting over the past few months.  
To the best of my knowledge, all of the cardholders that have a card are now designed as such, and, with the exception of a few files that Josh and I are going to address in special collections next week, the borrowing events are correctly assigned.  Once these final files are addressed, I'd feel comfortable signing off that this is as accurate representation of the data, both for library purposes and for eventual display on the website.

> I did notice two interesting issues that may need to be resolved with the export algorithm, though.

> 1.  If the footnote URL was incorrectly inputted (usually by not ending with .jp2, either ending with nothing or with .jpg), no images come up at all.  Is it possible to revise the script so that it can parse just the first part of the URL, then programmatically add the correct file ending? For example, just taking /r/roditi/00000001 and then adding the .jp2 to the end.
> 2.  Many of the cardholders are listed as having several of the same image (/0000001.jp2 over and over again, for example), instead of the correct progression of images.  I have attached a list of those files to this document.  
> My hypothesis is that somehow the export algorithm is having difficulty handling files that contain blank "surfaces," as many of the examples (sorted into "two-image" instances) have a blank verso.  The more complicated, multi-image instances all have a blank surface in there somewhere.
However, I have managed to find two-image instances with blank versos (Lewissohn and Levy, for example), that are handled correctly by the export.  The xml, to my eye, looks identical to the instances listed in the attached document, yet they were handled correctly. 
 

[Export image issues.txt](https://github.com/Princeton-CDH/mep-django/files/3045210/Export.image.issues.txt)

**Comment by rlskoeser** [comment]
@jkotin @elspethgreen @clmahoney : documenting what I've done so far in response to what Ellie found.  Thanks again @elspethgreen for the amazing attention to detail catching these problems.

- Adjusted the output to add `.jp2` to images that don't have an extension (I didn't see any ending with `.jpg` so I didn't add any logic for that)
- Ellie correctly identified a problem in the logic that was generating wrong output; I thought the database query I was using should have made the list of images distinct, but it didn't. I've adjusted the logic to fix that. This has corrected many of the errors Ellie identified in her attached file; I'll include notes on that below.
- When I revisited the export script I noticed I was only using footnotes from borrowing events, but we now have other events which could have footnotes (generic events and purchases), so I've revised the logic to look for any of those events with footnotes and use whichever image location it finds.
- A few of the people in the CSV output have no images at all. I checked Bryher and Brooks and I see no footnotes and no images in the card bibliography notes. Can someone check to see what we're missing on these?
- I saw at least one person in the output who has no images because the lending cards are at a different institution. I need to exclude those, but not sure the best way since there seems to be some slight inconsistency in the card bibliographies that have been manually added. It looks like maybe I could filter on "Sylvia Beach Papers" in the bibliographic note?

I can't remember if I shared the logic I'm currently using to generate the image list (if I did it was a while ago). I'm getting a list of all unique image locations for a member from footnotes associated with borrowing, purchasing, or generic events, ordered by the start date of the event; if that list is longer or equal to the length of images in the note of the bibliographic entry for the card, I use it. Otherwise I use the list in the card bibliography. This was the only way I could figure out to handle empty images that aren't otherwise tracked. This way we get the order from the events if possible, but should alway include all documented images.

I also thought of another way to sanity check this output: I'm going to ask the PUL devs if I can get a list of all the image files they have that should be included and then check them against the images in the CSV I'm generating.  Let me know if you're aware of any problems with that.

Here are my notes on the export image issues Ellie identified (see the attachment on the previous comment above:

- Two-image instances (recto has information, verso blank): all of these are now fixed except for Milward, which has the wrong information in the card bibliography note
- ✓  Georgette Camille (final image correctly listed as 00000003.jp2, but second [empty] image listed as 00000001.jp2): now listing as 1,2,3,4, I think this is correct
- ✓ Fourcade (blanks in s2 and s4) now listing as  1,2,3,4
- Goodwin (listed as three s4s followed by s3) (nonstandard content in s1 and s2, s4 for Enfield) - only listing 1,2,3,4 - simple fix is to edit card bibliography
- Kalbfleisch (s1 and s2 correct, s1 and s2 of Rhys affiliated with Kalbfleisch)  Because of mistake in header on Rhys card -- will fix - still only showing 1, 2; correct in card bibliography note?
- ✓ De Lanux (blank in s2)   now 1,2,3,4
- ✓ Mathews (blanks in s2, s4, and s6)   Checked database and images--borrowing events correctly assigned to s1 and s3 - now showing 1,2,3,4,5,6
- De Montricher (listed as s3, s2, s3, s4 etc.) (blank in s2) 3,2,3,4,5,6 - this is incorrect in card bibliography
- Oppen (listed as four s1’s) (only two sides, blank in s2) open/1,open/2,oppen/1,oppen/2 - this is incorrect in card bibliography
- Rice (listed as only s1) (two sides, both with content)  01 - only 01 in card bibliography; only 01 referenced in event footnotes; add 02 to card bibliography
- Rivoallan (blank in s3 and s4) repeating 01 over and over - based on bad list in card bibliography
- Wilde (blanks in s6 and s8) - repeating 01 over and over - based on bad list in card bibliography

**Comment by rlskoeser** [comment]
Worked with @clmahoney to clean up images and accounts based on the discrepancies we found when comparing to the PUL image list. Generated an export and checked the image list; only the expected 10 duplicate images (card images referencing more than one account) and briggs 3-130, which is a book that was digitized but not relevant for us. Cate and I both reviewed the CSV export, and then I shared with PUL collaborators so the next step can move forward.
